posthog/models/funnel.py

```diff
@@ -99,33 +86,111 @@ def _serialize_step(
             "type": step.type,
         }
 
-    def get_steps(self) -> List[Dict[str, Any]]:
-        filter = Filter(data=self.filters)
-        people = (
-            Person.objects.all()
-            .filter(team_id=self.team_id, persondistinctid__distinct_id__isnull=False)
-            .annotate(
-                **self._annotate_steps(
-                    team_id=self.team_id,
-                    filter=filter
+    def _build_query(self, query_bodies: dict):
+        """Build query using lateral joins using a combination of Django generated SQL
+           and sql built using psycopg2
+        """
+
+        ON_TRUE = "ON TRUE"
+        LEFT_JOIN_LATERAL = "LEFT JOIN LATERAL"
+        QUERY_HEADER = "SELECT {people}, {fields} FROM "
+        LAT_JOIN_BODY = """({query}) {step} {on_true} {join}"""
+        PERSON_FIELDS = [[sql.Identifier("posthog_person"), sql.Identifier("id")],
+                         [sql.Identifier("posthog_person"), sql.Identifier("created_at")],
+                         [sql.Identifier("posthog_person"), sql.Identifier("team_id")],
+                         [sql.Identifier("posthog_person"), sql.Identifier("properties")],
+                         [sql.Identifier("posthog_person"), sql.Identifier("is_user_id")]]
+        QUERY_FOOTER = sql.SQL("""
+            JOIN posthog_person ON posthog_person.id = {step0}.person_id
+            WHERE {step0}.person_id IS NOT NULL
+            GROUP BY {group_by}""")
+
+        person_fields = sql.SQL(",").join([sql.SQL(".").join(col) for col in PERSON_FIELDS])
+
+        steps = [sql.Identifier(step) for step, query in query_bodies.items()]
+        select_steps = [sql.Composed([
+            sql.SQL("MIN("),
+            step,
+            sql.SQL("."),
+            sql.Identifier("step_ts"),
+            sql.SQL(") as "),
+            step
+        ]) for step in steps]
+        lateral_joins = []
+        with connection.cursor() as cursor:
+            query_bodies = {step: cursor.mogrify(*q.query.sql_with_params()) for step, q in query_bodies.items()}
+        query = sql.SQL(QUERY_HEADER).format(
+            # step=steps[0],
+            fields=sql.SQL(',').join(select_steps),
+            people=person_fields
+        )
+        i = 0
+        for step, qb in query_bodies.items():
+            q = sql.SQL(qb.decode('utf-8')
+                        .replace("'1234321'", "{prev_step_person_id}")
```

Comment:
Is there a reason this couldn't happen in `_gen_lateral_bodies`? There might be, though if so might be good to add comment explaining that

Reply:
Good point. Moved to `_gen_lateral_bodies`